ddl/column.go

```diff
@@ -647,36 +648,39 @@ func onSetDefaultValue(t *meta.Meta, job *model.Job) (ver int64, _ error) {
 func needChangeColumnData(oldCol, newCol *model.ColumnInfo) bool {
 	toUnsigned := mysql.HasUnsignedFlag(newCol.Flag)
 	originUnsigned := mysql.HasUnsignedFlag(oldCol.Flag)
-	if oldCol.Tp == newCol.Tp && oldCol.Tp == mysql.TypeNewDecimal {
-		// Since type decimal will encode the precision, frac, negative(signed) and wordBuf into storage together, there is no short
-		// cut to eliminate data reorg change for column type change between decimal.
-		return oldCol.Flen != newCol.Flen || oldCol.Decimal != newCol.Decimal || toUnsigned != originUnsigned
-	}
-	if oldCol.Tp == newCol.Tp && (oldCol.Tp == mysql.TypeEnum || oldCol.Tp == mysql.TypeSet) {
-		return isElemsChangedToModifyColumn(oldCol.Elems, newCol.Elems)
-	}
-	if oldCol.Tp == mysql.TypeEnum || oldCol.Tp == mysql.TypeSet ||
-		newCol.Tp == mysql.TypeEnum || newCol.Tp == mysql.TypeSet {
-		return true
+	needTruncationOrToggleSign := func() bool {
+		return newCol.Flen > 0 && newCol.Flen < oldCol.Flen || toUnsigned != originUnsigned
 	}
 
-	switch oldCol.Tp {
-	case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp, mysql.TypeDuration, mysql.TypeYear:
-		switch newCol.Tp {
-		case mysql.TypeDate, mysql.TypeDatetime, mysql.TypeTimestamp, mysql.TypeDuration, mysql.TypeYear:
-			return oldCol.Tp != newCol.Tp
+	// deal with same type
```

Comment:
```suggestion
	// Deal with the same type.
```
the below comment should also be paid with attention.